File: src/main/java/fortress/tfol/FuncDecl.java
package fortress.tfol;

import fortress.data.ImmutableList;
import fortress.data.ImmutableWrapperList;
import java.util.List;
import java.util.ArrayList;
import fortress.util.Errors;

public class FuncDecl {
    private String name;
    private ImmutableList<Type> argTypes;
    private Type resultType;
    
    private FuncDecl(String name, ImmutableList<Type> argTypes, Type resultType) {
        Errors.precondition(argTypes.size() > 0, "Cannot create nullary functions; use a constant instead");
        Errors.precondition(! Names.isIllegal(name), "Illegal function name " + name);
        Errors.precondition(name.length() > 0, "Cannot create function with empty name");
        
        this.name = name;
        this.argTypes = argTypes;
        this.resultType = resultType;
    }
    
    public static FuncDecl mkFuncDecl(String name, List<Type> argTypes, Type resultType) {
        return new FuncDecl(name, ImmutableWrapperList.copyCollection(argTypes), resultType);
    }
    
    public static FuncDecl mkFuncDecl(String name, Type... types) {
        List<Type> argTypes = new ArrayList<>();
        Type resultType = types[types.length - 1];
        for(int i = 0; i < types.length - 1; i++) {
            argTypes.add(types[i]);
        }
        return mkFuncDecl(name, argTypes, resultType);
    }
    
    public int getArity() {
        return argTypes.size();
    }
    
    public String getName() {
        return name;
    }
    
    public List<Type> getArgTypes() {
        return argTypes;
    }
    
    public Type getResultType() {
        return resultType;
    }
    
    @Override
    public boolean equals(Object other) {
        if(this == other) {
            return true;
        }
        if(null == other) {
            return false;
        }
        if(this.getClass() != other.getClass()) {
            return false;
        }
        FuncDecl o = (FuncDecl) other;
        return this.name.equals(o.name)
            && this.argTypes.equals(o.argTypes)
            && this.resultType.equals(o.resultType);
    }
    
    @Override
    public int hashCode() {
        int prime = 31;
        int result = 1;
        result = result * prime + name.hashCode();
        result = result * prime + argTypes.hashCode();
        result = result * prime + resultType.hashCode();
        return result;
    }
}

File: src/main/scala/fortress/tfol/FuncDecl.scala
package fortress.tfol

import fortress.util.Errors
import scala.collection.JavaConverters._
import scala.collection.immutable.Seq // Default to immutable Seqs
import scala.annotation.varargs // So we can call Scala varargs methods from Java

case class FuncDecl(name: String, argTypes: Seq[Type], resultType: Type) {
    Errors.precondition(argTypes.size > 0, "Cannot create nullary functions; use a constant instead")
    Errors.precondition(! Names.isIllegal(name), "Illegal function name " + name)
    Errors.precondition(name.length > 0, "Cannot create function with empty name")
    
    def getArity: Int = argTypes.size
    
    def getName: String = name
    
    def getArgTypes: java.util.List[Type] = argTypes.asJava
    
    def getResultType: Type = resultType
    
    override def toString: String = name + ": (" + argTypes.mkString(", ") + ") -> " + resultType.toString
}

object FuncDecl {
    def mkFuncDecl(name: String, argTypes: java.util.List[Type], resultType: Type): FuncDecl =
        FuncDecl(name, argTypes.asScala.toList, resultType)
    
    @varargs
    def mkFuncDecl(name: String, types: Type*): FuncDecl = {
        val argTypes = new scala.collection.mutable.ListBuffer[Type]
        val resultType = types(types.size - 1);
        for(i <- 0 to (types.size - 2)) {
            argTypes += types(i)
        }
        FuncDecl(name, argTypes.toList, resultType);
    }
}
